### PR TITLE
DSD-418: Chakra/DS Button component

### PR DESCRIPTION
**Pull request by EdwinGuzman**
Fixes JIRA ticket [DSD-418](https://jira.nypl.org/browse/DSD-418)

## This PR does the following:
- Uses Chakra's `Button` component to compose the DS `Button` component.
- This is a WIP but setting the PR to view the preview Storybook instance.

### Checklist:

- [x] I have updated the Storybook documentation accordingly.
- [x] All new and existing tests passed.

### Front End Review:

- [x] View [the example in Storybook]()

### CODE REVIEW

**Comment by EdwinGuzman** [comment]
Update from meeting:
- we'll remove SCSS and move forward with css-in-js
- there's going to be a variant for the red donate button style
- We need to work on the global styles and themes for spacing and other rules ASAP
- I'm going to stick with `buttonType` over `variant` for the prop type (for now)
- for a separate PR we'll document how to write style props and what props are available for DS component and why there's a need to restrict adding style props to Chakra components that are re-exported by DS

I'll continue working on this PR based on the above.

@wluisi @hokei @bigfishdesign13

**Comment by hokei** [comment]
> there's going to be a variant for the red donate button style

Does that mean we're purposely not exposing the other props, or does that mean we're still going to expose them but make it clear in the docs that those other props should be used as a last resort?  @bigfishdesign13 

(PS we don't have to make clear what the chakra props are in the stories, we can just link to Chakra and be like, "these are available but use them as a last resort")

**Comment by EdwinGuzman** [comment]
> Does that mean we're purposely not exposing the other props, or does that mean we're still going to expose them but make it clear in the docs that those other props should be used as a last resort? @bigfishdesign13

For now, none of the style props are actually available for the DS `Button` component since that's our custom component. In this case, we're not exposing the other props but it's not intentional. We just haven't fully decided on whether we should use a `{...rest}` approach for any and all other props or keep using the attribute. Maybe we can hold off on that decision until more components are built and we see, through implementation, what the best approach is.

> (PS we don't have to make clear what the chakra props are in the stories, we can just link to Chakra and be like, "these are available but use them as a last resort")

I'm already doing this (except for the "use them as a last resort" message) in the other PR the re-exports Chakra components directly from the DS - https://pr640-1cyzfp5o9mouqwm0dsk0lxabk2hgoxgl.tugboat.qa/?path=/docs/chakra-layout-box--box
There's a link to the style props documentation in the Story above.

**Comment by EdwinGuzman** [comment]
This is no longer a WIP and is ready for a full review.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr642-0kvfxcldirptzwjy7ykojz9esatodof9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/611d668c331d463cd063d4de

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr642-0kvfxcldirptzwjy7ykojz9esatodof9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/611d668c331d463cd063d4de

**Review by hokei — APPROVED** [review]
This looks good to me!  Most of my questions are around CSS going forward, which seems out of scope for this PR.

File: src/components/Button/Button.tsx
Comment (nypl-wluisi) [inline src/components/Button/Button.tsx:90]: Are we keeping `className` and bem related around temporarily for backwards compatibility reasons?

**Comment by nypl-wluisi** [comment]
This might be pre-existing, but the naming of the props `buttonType` and `type` is a bit confusing. `buttonType` prop controls visual changes for the component itself, where as `type` sets the actual `type` attribute on the button HTML element?

**Comment by nypl-wluisi** [comment]
Search Bar component has some styling issues, related to this button change: https://pr642-0kvfxcldirptzwjy7ykojz9esatodof9.tugboat.qa/?path=/story/components-form-elements-searchbar--basic

search-bar-bug

**Comment by hokei** [comment]
> Search Bar component has some styling issues, related to this button change: 

This is the kind of thing I'd love to figure out if we can catch using automated tests - would a snapshot catch this?

File: src/components/Button/_Button.scss
Comment (bigfishdesign13) [inline src/components/Button/_Button.scss:8]: What is necessitating the `!important`?

File: src/theme/components/button.ts
Comment (bigfishdesign13) [inline src/theme/components/button.ts]: This is something that is dependent on another PR, but should you replace the `space` CSS vars with the Chakra data objects before merging this?

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr642-0kvfxcldirptzwjy7ykojz9esatodof9.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/611d668c331d463cd063d4de

**Review by bigfishdesign13 — COMMENTED** [review]
I think this looks good as we start shifting away from the SCSS.  I just have some minor questions.

File: src/components/Button/Button.tsx
Comment (EdwinGuzman) [reply src/components/Button/Button.tsx:90]: Yes, and also because we haven't completely moved away from bem so maybe this should stick around until we no longer need it.

**Comment by EdwinGuzman** [comment]
> This might be pre-existing, but the naming of the props buttonType and type is a bit confusing. buttonType prop controls visual changes for the component itself, where as type sets the actual type attribute on the button HTML element?

Yeah, it's confusing but you're right about how it works.

**Comment by EdwinGuzman** [comment]
> This is the kind of thing I'd love to figure out if we can catch using automated tests - would a snapshot catch this?

Yes, exactly. We just now need to implement more snapshot tests in order to catch these issues.

File: src/components/Button/_Button.scss
Comment (EdwinGuzman) [reply src/components/Button/_Button.scss:8]: This is only really being used in the `Accordion` component to style the input. Some of Chakra's inline css rules take precedence so it needs the `!important`. But once we update the `Accordion` component, this can be removed.

File: src/theme/components/button.ts
Comment (EdwinGuzman) [reply src/theme/components/button.ts]: This is fixed in #648 so I'll leave it there to see the progression of the Chakra integration for this component.

File: src/components/Button/Button.tsx
Comment (nypl-wluisi) [reply src/components/Button/Button.tsx:90]: Should we consider adding a deprecation flag comment for this and other parts that we're planning on getting rid of? Something like:

```js
// @deprecated since version 2.3.
// Will be deleted in version 3.0.
 ```

Comment (bigfishdesign13) [reply src/components/Button/Button.tsx:90]: @nypl-wluisi Doing this on a code level seems untenable.  If we are going to flag things like this, what about adding a `Deprecated` page to the `Documentation` section in Storybook?

Comment (nypl-wluisi) [reply src/components/Button/Button.tsx:90]: Yeah, that's a good point. We can discuss more, but we need a list somewhere of all props and other things that we plan on no longer using with the switch to Chakra. And then we also need to decide if something like `bem` and `className` are things we are going to move away from, which I think we are?

Comment (EdwinGuzman) [reply src/components/Button/Button.tsx:90]: Yes, I think we should move away from `bem` with this Chakra update but it'll be a slow transition. I do like that the current function with `className={bem(baseClass, [], "", [className])}` (where classname = "newclass") will render `"button newclass". We might have to replace this with 
```
className={`${baseClass} ${className}`}
```

**Comment by EdwinGuzman** [comment]
Snapshot testing catching visual differences:

I reverted the Button component, added a snapshot test for the `SearchBar` component, un-reverted the Button changes, and then ran the tests again and it did fail:

Screen Shot 2021-08-25 at 3 40 27 PM

Ignore the second error with the div/button -- this one fails because the `id` is autogenerated and it will always be different and we need to fix this separately. But you can see that the `button` component has new classes so we can infer that maybe it also changed visually. This is helpful but we need to set time aside first to write snapshot tests in order to catch more of these issues in the future. I need to revert this snapshot test since the `Icon` component is its own issue, but I'll fix the button so it looks right in the `SearchBar` component.

**Comment by EdwinGuzman** [comment]
Updated.

**Comment by EdwinGuzman** [comment]
@wluisi  the latest commit should fix the svg issue in the header nav buttons, but it's global so it should fix it everywhere else.

**Pull request merged**
